src/AlgoliaSearchCore.js

```diff
@@ -438,7 +421,62 @@ AlgoliaSearchCore.prototype._jsonRequest = function(initialOpts) {
     }
   }
 
-  var promise = doRequest(
+  function isCacheValidWithCurrentID() {
+    return client._useCache && cache && cache[cacheID] !== undefined;
+  }
+
+  function interopCallbackReturn(request, callback) {
+    request.catch(function() {
+      // Release the cache on error
+      if (isCacheValidWithCurrentID()) {
+        delete cache[cacheID];
+      }
+    });
+
+    // either we have a callback
+    // either we are using promises
+    if (typeof initialOpts.callback === 'function') {
+      request.then(function okCb(content) {
+        exitPromise(function() {
+          initialOpts.callback(null, callback(content));
+        }, client._setTimeout || setTimeout);
+      }, function nookCb(err) {
+        exitPromise(function() {
+          initialOpts.callback(err);
+        }, client._setTimeout || setTimeout);
+      });
+    } else {
+      return request.then(callback);
+    }
+  }
+
+  if (client._useCache) {
+    cacheID = initialOpts.url;
+  }
+
+  // as we sometime use POST requests to pass parameters (like query='aa'),
+  // the cacheID must also include the body to be different between calls
+  if (client._useCache && body) {
+    cacheID += '_body_' + body;
+  }
+
+  if (isCacheValidWithCurrentID()) {
+    requestDebug('serving response from cache');
+
+    var maybePromiseForCache = cache[cacheID];
+
+    // In case the cache is warmup with value that is not a promise
```

Comment:
When does that happens?

Reply:
when a user passes `algoliasearch('', '', { cache: myWarmedUpCacheWhichIsObjectsNotPromise });` I think